fix(ivy): trigger directive inputs with i18n translations

ocombe opened the pull request.

mhevery commented on packages/core/src/render3/i18n.ts:
```diff
@@ -879,6 +877,16 @@ function readUpdateOpCodes(
+    nodeIndex: number, attrName: string, attrValue: string, viewData: LView,
+    sanitizeFn?: SanitizerFn | null) {
+  const inputData = initializeTNodeInputs(getTNode(nodeIndex, viewData));
+  if (inputData && inputData[attrName]) {
```
`inputData[attrName]` is a megamorphic read so we should solve it in a different way.

ocombe replied:
we use this a few times in shared.ts

ocombe replied:
Any idea how we could do it any other way?

mhevery commented on packages/core/src/render3/i18n.ts:
```diff
@@ -879,6 +877,16 @@ function readUpdateOpCodes(
+  if (inputData && inputData[attrName]) {
+    ɵɵelementProperty(nodeIndex, attrName, attrValue, sanitizeFn);
+  }
+  ɵɵelementAttribute(nodeIndex, attrName, attrValue, sanitizeFn);
```
I think at this point instead of calling the instructions you should just directly call either `setAttribute` or `comp[prop] = attrValue`. The instruction should be reserved for template function.

Calling `elementProperty` or `elementAttribute` is wrong because it will allocate space in `LView` data

ocombe replied:
when we use i18n we remove those instructions from the template and let i18n handle those attributes (if I'm not mistaken). Not calling those would probably make things fails

ocombe replied:
I checked and `ɵɵelementAttribute` doesn't allocate anything in `LView`, it only sets the attribute on the element.
But that is true for `ɵɵelementProperty`. That being said we need to call that function in case of directives, just setting the property is not enough to trigger the inputs.

pkozlowski-opensource replied:
I would _very_ much prefer (actually, I would love to enforce it as a rule in the ngIvy code-base!) where we do NOT call instructions from other instructions (and call them from the generated code only).

If you need part of the  `ɵɵelementProperty` logic just please refactor this instruction by pulling common code into a new function. Calling one instruction from another instruction means that we need to look-up global state (LView, indexes etc.) multiple times and this is not great for perf and general code understanding.

ocombe commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -538,8 +539,13 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
+          // We treat this attribute as if it was a binding so that we don't risk calling inputs
+          // with the untranslated value.
+          // Also this generates smaller templates until FW-1248 is fixed.
+          allOtherInputs.push(attr);
```
I agree but I didn't want to mess with those things because I don't know enough about it. I'm not sure what I'll have to change if I add a new symbol for that?

kara replied:
Maybe let's add a JIRA to clean it up plus a TODO comment referencing it here?

ocombe commented on packages/core/src/render3/i18n.ts:
```diff
@@ -742,7 +740,8 @@ function readCreateOpCodes(
           const attrValue = createOpCodes[++i] as string;
-          ɵɵelementAttribute(elementNodeIndex, attrName, attrValue);
+          elementPropertyInternal(elementNodeIndex, attrName, attrValue);
+          elementAttributeInternal(elementNodeIndex, attrName, attrValue);
```
sorry this was a mistake, good catch

ocombe commented on packages/core/src/render3/i18n.ts:
```diff
@@ -961,14 +960,23 @@ function i18nAttributesFirstPass(tView: TView, index: number, values: string[])
+          const tNode = getTNode(previousElementIndex, lView);
+          let inputData: PropertyAliases|null|undefined;
+          let dataValue: PropertyAliasValue|undefined;
+          if ((inputData = initializeTNodeInputs(tNode)) && (dataValue = inputData[attrName])) {
```
good point, thanks

mhevery commented on packages/core/src/render3/instructions/element.ts:
```diff
@@ -206,27 +206,32 @@ export function ɵɵelementAttribute(
+export function elementAttributeInternal(
+    index: number, name: string, value: any, sanitizer?: SanitizerFn | null, namespace?: string) {
+  ngDevMode && validateAgainstEventAttributes(name);
+  const lView = getLView();
```
pull `lView` and `renderer` into `ɵɵelementAttribute`.

per our discussion in the meeting only instructions should be able to access global data.

mhevery commented on packages/core/test/render3/render_util.ts:
```diff
@@ -147,12 +148,14 @@ export class ComponentFixture<T> extends BaseFixture {
+        rendererFactory?: RendererFactory3,
+        playerHandler?: PlayerHandler
+      } = {},
+      public keepNgReflect = false) {
```
Why is `keepNgReflect` not part of the options?

mhevery commented on packages/core/src/render3/i18n.ts:
```diff
@@ -7,20 +7,18 @@
 import {LContainer, NATIVE} from './interfaces/container';
 import {COMMENT_MARKER, ELEMENT_MARKER, I18nMutateOpCode, I18nMutateOpCodes, I18nUpdateOpCode, I18nUpdateOpCodes, IcuType, TI18n, TIcu} from './interfaces/i18n';
-import {TElementNode, TIcuContainerNode, TNode, TNodeType} from './interfaces/node';
+import {PropertyAliasValue, PropertyAliases, TElementNode, TIcuContainerNode, TNode, TNodeType} from './interfaces/node';
```
clean up unneeded imports

mhevery reviewed (approved):
LGTM, with cleanup

kara reviewed (approved):
LGTM once remaining comments are addressed (Misko's review and adding the TODO). 

Also move this out of draft?

ocombe committed:
fix(ivy): trigger directive inputs with i18n translations

Changed runtime i18n to define attributes with bindings, or matching directive inputs/outputs as element properties as we are supposed to do in Angular.
This PR fixes the issue where directive inputs wouldn't be trigged.

FW-1315 #resolve

ocombe committed:
feat(ivy): add `AttributeMarker.I18n` for i18n attributes

`i18nAttributes` instructions always occur after the element instruction. This means that we need to treat `i18n-` attributes differently.
By defining a specific `AttributeMarker` we can ensure that we won't trigger directive inputs with untranslated attribute values.

FW-1332 #resolve

kara commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -1120,9 +1117,9 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
    * because those values are intended to always be generated as property instructions.
```
Nit: update comment above to include I18N markers?

kara reviewed (approved):
LGTM

kara commented:
[presubmit](http://test/OCL:248251134:BASE:248775226:1558124190537:cf38633b)

alxhub reviewed (approved):
👍 for `fw/compiler`

matsko commented:
presubmit = https://tap.corp.google.com/ui#id=OCL:249864181:BASE:249864229:1558721339997:42331cd8

matsko commented:
@ocombe rebase please.

alxhub commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -538,8 +538,9 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
           // TODO(FW-1248): prevent attributes duplication in `i18nAttributes` and `elementStart`
           // arguments
           i18nAttrs.push(attr);
+        } else {
```
Is the above comment now outdated, as the attributes are not kept in the main list?

matsko commented:
https://test.corp.google.com/ui#id=OCL:249864181:BASE:250743369:1559245977908:f087d8d5
